File: team2/sbm-restaurant-app/backend/src/main/java/com/cs_25_2_team2/RestaurantManagementApp/services/KitchenService.java
package com.cs_25_2_team2.RestaurantManagementApp.services;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.springframework.stereotype.Service;

import com.cs_25_2_team2.RestaurantManagementApp.Chef;
import com.cs_25_2_team2.RestaurantManagementApp.Order;
import com.cs_25_2_team2.RestaurantManagementApp.OrderQueue;
import com.cs_25_2_team2.RestaurantManagementApp.Staff;

/**
 * Service class for Kitchen management using the existing OrderQueue, Chef, and Staff backend classes.
 * Handles order assignment, kitchen workflow, and chef management.
 * 
 * Uses the existing backend architecture:
 * - OrderQueue class for order management
 * - Chef class for chef operations (extends Staff)
 * - Staff class hierarchy for kitchen staff
 * - Order class with status management
 * 
 * @author Team 2
 * @version 1.0
 */
@Service
public class KitchenService {
    
    private final OrderQueue orderQueue;
    private final Map<String, Chef> chefs = new HashMap<>();
    // Use Long for chef keys for consistency
    private final Map<Long, Chef> chefsById = new HashMap<>();
    private final Map<String, Staff> kitchenStaff = new HashMap<>();
    
    /**
     * Constructor initializes kitchen with sample staff
     */
    public KitchenService() {
        this.orderQueue = new OrderQueue();
        initializeSampleKitchenStaff();
    }
    
    /**
     * Get all pending orders for kitchen preparation
     */
    public List<Order> getPendingOrders() {
        // Return orders with status Pending
        return orderQueue.getAll().stream()
            .filter(order -> order.getStatus() == Order.Status.Pending)
            .toList();
    }
    
    /**
     * Get orders currently being prepared
     */
    public List<Order> getOrdersInPreparation() {
        // Return orders with status Preparing
        return orderQueue.getAll().stream()
            .filter(order -> order.getStatus() == Order.Status.Preparing)
            .toList();
    }
    
    /**
     * Get completed orders ready for delivery
     */
    public List<Order> getReadyOrders() {
        // Return orders with status ReadyForDelivery
        return orderQueue.getAll().stream()
            .filter(order -> order.getStatus() == Order.Status.ReadyForDelivery)
            .toList();
    }
    
    /**
     * Add order to kitchen queue
     */
    public void addOrderToQueue(Order order) {
    orderQueue.add(order);
    System.out.println("Order queued: " + order.toString());
    }
    
    /**
     * Start preparing an order - assigns to available chef
     */
    public boolean startPreparingOrder(Long orderId) {
        Order order = findOrderById(orderId);
        if (order == null) {
            return false;
        }
        
        // Find available chef
        Chef availableChef = findAvailableChef();
        if (availableChef != null) {
            try {
                // Use existing Chef.assignOrder() method
                availableChef.assignOrder(order);
                
                // Update order status using existing Order.updateStatus() method
                order.updateStatus(Order.Status.Preparing);
                
                return true;
            } catch (Exception e) {
                System.err.println("Error starting order preparation: " + e.getMessage());
                return false;
            }
        }
        
        return false; // No available chef
    }
    
    /**
     * Mark order as ready/complete
     */
    public boolean completeOrder(Long orderId) {
        Order order = findOrderById(orderId);
        if (order == null) {
            return false;
        }
        
        try {
            // Update order status using existing Order.updateStatus() method
            order.updateStatus(Order.Status.ReadyForDelivery);
            
            // Find the chef who was preparing this order and mark them as available
            Chef chef = findChefByOrder(order);
            if (chef != null) {
                // The chef's assigned orders will be automatically updated
                // since we're using the existing backend architecture
            }
            
            return true;
        } catch (Exception e) {
            System.err.println("Error completing order: " + e.getMessage());
            return false;
        }
    }
    
    /**
     * Get the complete order queue status
     */
    public OrderQueue getOrderQueue() {
        return orderQueue;
    }
    
    /**
     * Get all chefs in the kitchen
     */
    public List<Chef> getAllChefs() {
    return new ArrayList<>(chefsById.values());
    }
    
    /**
     * Get all kitchen staff
     */
    public List<Staff> getAllKitchenStaff() {
        return new ArrayList<>(kitchenStaff.values());
    }
    
    /**
     * Get chef by ID
     */
    public Chef getChefById(Long chefId) {
    return chefsById.get(chefId);
    }
    
    /**
     * Add chef to kitchen
     */
    public void addChef(Chef chef) {
    chefsById.put(chef.getRawId(), chef);
    kitchenStaff.put(String.valueOf(chef.getRawId()), chef);
    }
    
    /**
     * Get available chefs (not busy)
     */
    public List<Chef> getAvailableChefs() {
        return chefsById.values().stream()
            .filter(chef -> !chef.isBusy())
            .toList();
    }
    
    /**
     * Get orders assigned to a specific chef
     */
    public List<Order> getOrdersForChef(String chefId) {
    Chef chef = chefsById.get(Long.valueOf(chefId));
    return chef != null ? chef.getAssignedOrders() : new ArrayList<>();
    }
    
    /**
     * Estimate preparation time for current queue
     */
    public int estimatePreparationTime() {
        int pendingOrders = getPendingOrders().size();
        int preparingOrders = getOrdersInPreparation().size();
        int availableChefs = getAvailableChefs().size();
        
        if (availableChefs == 0) {
            return -1; // No chefs available
        }
        
        // Simple estimation: 10 minutes per order, adjusted for available chefs
        int totalOrders = pendingOrders + preparingOrders;
        return (totalOrders * 10) / Math.max(1, availableChefs);
    }
    
    /**
     * Find order by ID across all queue statuses
     */
    private Order findOrderById(Long orderId) {
        // Check pending orders
        for (Order order : getPendingOrders()) {
            if (order.getId() == orderId.intValue()) {
                return order;
            }
        }
        
        // Check orders in preparation
        for (Order order : getOrdersInPreparation()) {
            if (order.getId() == orderId.intValue()) {
                return order;
            }
        }
        
        // Check ready orders
        for (Order order : getReadyOrders()) {
            if (order.getId() == orderId.intValue()) {
                return order;
            }
        }
        
        return null;
    }
    
    /**
     * Find available chef using existing Chef.isBusy() method
     */
    private Chef findAvailableChef() {
        return chefsById.values().stream()
            .filter(chef -> !chef.isBusy())
            .findFirst()
            .orElse(null);
    }
    
    /**
     * Find chef assigned to a specific order
     */
    private Chef findChefByOrder(Order order) {
        return chefsById.values().stream()
            .filter(chef -> chef.getAssignedOrders().contains(order))
            .findFirst()
            .orElse(null);
    }
    
    /**
     * Initialize kitchen with sample chefs and staff
     */
    private void initializeSampleKitchenStaff() {
        // Sample chefs
        Chef chef1 = new Chef("Gordon Ramsay", "456 Kitchen Ave", "987-654-3210", 1L);
        Chef chef2 = new Chef("Julia Child", "789 Culinary St", "555-987-6543", 2L);
        Chef chef3 = new Chef("Anthony Bourdain", "321 Food Blvd", "444-555-6666", 3L);

        addChef(chef1);
        addChef(chef2);
        addChef(chef3);
        
    System.out.println("Initialized kitchen with " + chefsById.size() + " chefs");
    }
    
    /**
     * Get kitchen statistics for admin dashboard
     */
    public Map<String, Object> getKitchenStatistics() {
        Map<String, Object> stats = new HashMap<>();
    stats.put("totalChefs", chefsById.size());
        stats.put("availableChefs", getAvailableChefs().size());
        stats.put("pendingOrders", getPendingOrders().size());
        stats.put("preparingOrders", getOrdersInPreparation().size());
        stats.put("readyOrders", getReadyOrders().size());
        stats.put("estimatedWaitTime", estimatePreparationTime());
        return stats;
    }
}

File: team2/sbm-restaurant-app/frontend/src/services/kitchenService.js
import axios from 'axios';

// Configure axios base URL to match your Spring Boot backend
const API_BASE_URL = 'http://localhost:8080/api';

// Create axios instance with base configuration
const api = axios.create({
  baseURL: API_BASE_URL,
  headers: {
    'Content-Type': 'application/json',
  },
  timeout: 10000,
});

// Mock kitchen data for development when backend is unavailable
const mockChefs = [
  {
    id: "CHEF001",
    name: "Gordon Ramsay",
    phoneNumber: "987-654-3210",
    address: "456 Kitchen Ave",
    role: "CHEF",
    isBusy: false,
    assignedOrdersCount: 0
  },
  {
    id: "CHEF002",
    name: "Julia Child",
    phoneNumber: "555-987-6543",
    address: "789 Culinary St",
    role: "CHEF",
    isBusy: true,
    assignedOrdersCount: 2
  },
  {
    id: "CHEF003",
    name: "Anthony Bourdain",
    phoneNumber: "444-555-6666",
    address: "321 Food Blvd",
    role: "CHEF",
    isBusy: false,
    assignedOrdersCount: 1
  }
];

const mockKitchenOrders = [
  {
    id: 101,
    customerId: 1,
    customerName: "John Doe",
    totalPrice: 13.97,
    status: "Preparing",
    createdAt: "2025-10-22T10:30:00",
    items: ["Texas Loaded Baked Potato x2", "Potato Salad x1"]
  },
  {
    id: 102,
    customerId: 1,
    customerName: "John Doe",
    totalPrice: 4.99,
    status: "ReadyForDelivery",
    createdAt: "2025-10-22T12:45:00",
    items: ["Loaded Baked Potato Soup x1"]
  },
  {
    id: 103,
    customerId: 2,
    customerName: "Jane Smith",
    totalPrice: 14.99,
    status: "Delivered",
    createdAt: "2025-10-22T11:20:00",
    items: ["Aloo Tikki Chaat x1"]
  }
];

// Kitchen API service functions
export const kitchenService = {
  // Get all pending orders for kitchen preparation
  getPendingOrders: async () => {
    try {
      const response = await api.get('/kitchen/orders/pending');
      return response.data;
    } catch (error) {
      console.warn('Backend API unreachable for getPendingOrders(), using mock data:', error.message);
      return mockKitchenOrders.filter(order => [
        'Pending',
        'Delivered',
        'Preparing',
        'Out for Delivery',
        'ReadyForDelivery',
        'Cancelled'
      ].includes(order.status));
    }
  },

  // Start preparing an order - assigns to available chef
  startPreparingOrder: async (orderId) => {
    try {
      const response = await api.put(`/kitchen/orders/${orderId}/start`);
      return response.data;
    } catch (error) {
      console.warn(`Backend API unreachable for startPreparingOrder(${orderId}), using mock implementation:`, error.message);
      
      const order = mockKitchenOrders.find(o => o.id === orderId);
      if (order && order.status === 'Pending') {
        order.status = 'Preparing';
        return {
          success: true,
          message: "Order preparation started successfully",
          orderId: orderId
        };
      }
      
      return {
        success: false,
        message: "Failed to start order preparation - no available chef or order not found"
      };
    }
  },

  // Mark order as ready/complete
  completeOrder: async (orderId) => {
    try {
      const response = await api.put(`/kitchen/orders/${orderId}/complete`);
      return response.data;
    } catch (error) {
      console.warn(`Backend API unreachable for completeOrder(${orderId}), using mock implementation:`, error.message);
      
      const order = mockKitchenOrders.find(o => o.id === orderId);
      if (order && order.status === 'Preparing') {
        order.status = 'Out for Delivery';
        return {
          success: true,
          message: "Order marked as out for delivery",
          orderId: orderId
        };
      }
      if (order && order.status === 'Out for Delivery') {
        order.status = 'ReadyForDelivery';
        return {
          success: true,
          message: "Order marked as ready for delivery",
          orderId: orderId
        };
      }
      
      return {
        success: false,
        message: "Failed to complete order - order not found or already completed"
      };
    }
  },

  // Get the complete kitchen order queue
  getOrderQueue: async () => {
    try {
      const response = await api.get('/kitchen/order-queue');
      return response.data;
    } catch (error) {
      console.warn('Backend API unreachable for getOrderQueue(), using mock data:', error.message);
      
      return {
        pending: mockKitchenOrders.filter(order => order.status === 'Pending'),
        delivered: mockKitchenOrders.filter(order => order.status === 'Delivered'),
        preparing: mockKitchenOrders.filter(order => order.status === 'Preparing'),
        outForDelivery: mockKitchenOrders.filter(order => order.status === 'Out for Delivery'),
        readyForDelivery: mockKitchenOrders.filter(order => order.status === 'ReadyForDelivery'),
        cancelled: mockKitchenOrders.filter(order => order.status === 'Cancelled')
      };
    }
  },

  // Get estimated preparation time
  getEstimatedPreparationTime: async () => {
    try {
      const response = await api.post('/kitchen/estimate-time');
      return response.data;
    } catch (error) {
      console.warn('Backend API unreachable for getEstimatedPreparationTime(), using mock data:', error.message);
      
  const pendingCount = mockKitchenOrders.filter(o => o.status === 'Pending').length;
  const deliveredCount = mockKitchenOrders.filter(o => o.status === 'Delivered').length;
  const preparingCount = mockKitchenOrders.filter(o => o.status === 'Preparing').length;
  const outForDeliveryCount = mockKitchenOrders.filter(o => o.status === 'Out for Delivery').length;
  const readyForDeliveryCount = mockKitchenOrders.filter(o => o.status === 'ReadyForDelivery').length;
  const cancelledCount = mockKitchenOrders.filter(o => o.status === 'Cancelled').length;
      const availableChefs = mockChefs.filter(chef => !chef.isBusy).length;
      
  const totalOrders = pendingCount + deliveredCount + preparingCount + outForDeliveryCount + readyForDeliveryCount + cancelledCount;
      const estimatedMinutes = availableChefs > 0 ? Math.ceil((totalOrders * 10) / availableChefs) : -1;
      
      return {
        estimatedMinutes,
        estimatedTime: estimatedMinutes > 0 ? `${estimatedMinutes} minutes` : "No chefs available",
        availableChefs,
        totalChefs: mockChefs.length
      };
    }
  },

  // Get all chefs in the kitchen
  getAllChefs: async () => {
    try {
      const response = await api.get('/kitchen/chefs');
      return response.data;
    } catch (error) {
      console.warn('Backend API unreachable for getAllChefs(), using mock data:', error.message);
      return mockChefs;
    }
  },

  // Get available chefs (not busy)
  getAvailableChefs: async () => {
    try {
      const response = await api.get('/kitchen/chefs/available');
      return response.data;
    } catch (error) {
      console.warn('Backend API unreachable for getAvailableChefs(), using mock data:', error.message);
      return mockChefs.filter(chef => !chef.isBusy);
    }
  },

  // Get orders assigned to a specific chef
  getOrdersForChef: async (chefId) => {
    try {
      const response = await api.get(`/kitchen/chefs/${chefId}/orders`);
      return response.data;
    } catch (error) {
      console.warn(`Backend API unreachable for getOrdersForChef(${chefId}), using mock data:`, error.message);
      
      // For mock data, return preparing orders for busy chefs
      const chef = mockChefs.find(c => c.id === chefId);
      if (chef && chef.isBusy) {
        return mockKitchenOrders.filter(order => order.status === 'Preparing').slice(0, chef.assignedOrdersCount);
      }
      return [];
    }
  },

  // Get kitchen statistics for admin dashboard
  getKitchenStatistics: async () => {
    try {
      const response = await api.get('/kitchen/statistics');
      return response.data;
    } catch (error) {
      console.warn('Backend API unreachable for getKitchenStatistics(), using mock data:', error.message);
      
      const availableChefs = mockChefs.filter(chef => !chef.isBusy).length;
      const estimatedTime = await kitchenService.getEstimatedPreparationTime();
      
      return {
        totalChefs: mockChefs.length,
        availableChefs,
        pendingOrders: mockKitchenOrders.filter(o => o.status === 'Pending').length,
        deliveredOrders: mockKitchenOrders.filter(o => o.status === 'Delivered').length,
        preparingOrders: mockKitchenOrders.filter(o => o.status === 'Preparing').length,
        outForDeliveryOrders: mockKitchenOrders.filter(o => o.status === 'Out for Delivery').length,
        readyForDeliveryOrders: mockKitchenOrders.filter(o => o.status === 'ReadyForDelivery').length,
        cancelledOrders: mockKitchenOrders.filter(o => o.status === 'Cancelled').length,
        estimatedWaitTime: estimatedTime.estimatedMinutes
      };
    }
  },

  // Add order to kitchen queue (called from order creation)
  addOrderToQueue: async (orderData) => {
    try {
      const response = await api.post('/kitchen/orders', orderData);
      return response.data;
    } catch (error) {
      console.warn('Backend API unreachable for addOrderToQueue(), using mock implementation:', error.message);
      
      // Add to mock orders
      const newOrder = {
        id: Math.floor(1000 + Math.random() * 9000),
        ...orderData,
        status: 'Pending',
        createdAt: new Date().toISOString()
      };
      
      mockKitchenOrders.push(newOrder);
      
      return {
        success: true,
        message: "Order added to kitchen queue"
      };
    }
  }
};

export default kitchenService;